test/Microsoft.AspNet.Server.KestrelTests/UrlPathDecoder.cs

```diff
@@ -0,0 +1,171 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Linq;
+using Microsoft.AspNet.Server.Kestrel.Http;
+using Microsoft.AspNet.Server.Kestrel.Infrastructure;
+using Xunit;
+
+namespace Microsoft.AspNet.Server.KestrelTests
+{
+    public class UrlPathDecoderTests
+    {
+
+        [Fact]
+        public void Empty()
+        {
+            PositiveAssert(string.Empty, string.Empty);
+        }
+
+        [Fact]
+        public void WhiteSpace()
+        {
+            PositiveAssert("    ", "    ");
+        }
+
+        [Theory]
+        [InlineData("/foo/bar", "/foo/bar")]
+        [InlineData("/foo/BAR", "/foo/BAR")]
+        [InlineData("/foo/", "/foo/")]
+        [InlineData("/", "/")]
+        public void NormalCases(string raw, string expect)
+        {
+            PositiveAssert(raw, expect);
+        }
+
+        [Theory]
+        [InlineData("%2F", "%2F")]
+        [InlineData("/foo%2Fbar", "/foo%2Fbar")]
+        [InlineData("/foo%2F%20bar", "/foo%2F bar")]
+        public void SkipForwardSlash(string raw, string expect)
+        {
+            PositiveAssert(raw, expect);
+        }
+
+        [Theory]
+        [InlineData("%D0%A4", "Ф")]
+        [InlineData("%d0%a4", "Ф")]
+        [InlineData("%E0%A4%AD", "भ")]
+        [InlineData("%e0%A4%Ad", "भ")]
+        [InlineData("%F0%A4%AD%A2", "𤭢")]
+        [InlineData("%F0%a4%Ad%a2", "𤭢")]
+        [InlineData("%48%65%6C%6C%6F%20%57%6F%72%6C%64", "Hello World")]
+        [InlineData("%48%65%6C%6C%6F%2D%C2%B5%40%C3%9F%C3%B6%C3%A4%C3%BC%C3%A0%C3%A1", "Hello-µ@ßöäüàá")]
+        // Overlong borderline cases
+        public void ValidUTF8(string raw, string expect)
+        {
+            PositiveAssert(raw, expect);
+        }
+
+        [Theory]
+        [InlineData("%C3%84ra%20Benetton", "Ära Benetton")]
+        [InlineData("%E6%88%91%E8%87%AA%E6%A8%AA%E5%88%80%E5%90%91%E5%A4%A9%E7%AC%91%E5%8E%BB%E7%95%99%E8%82%9D%E8%83%86%E4%B8%A4%E6%98%86%E4%BB%91", "我自横刀向天笑去留肝胆两昆仑")]
+        public void Internationalized(string raw, string expect)
+        {
+            PositiveAssert(raw, expect);
+        }
+
+        [Theory]
+        // Test the borderline cases of overlong UTF8.
+        [InlineData("%C2%80")]
+        [InlineData("%C2%80")]
+        [InlineData("%E0%A0%80")]
+        [InlineData("%F0%90%80%80")]
+        public void ValidUTF8(string raw)
+        {
+            PositiveAssert(raw);
+        }
+
+        [Theory]
+        // Overlong ASCII
+        [InlineData("%C0%A4")]
+        [InlineData("%C1%BF")]
+        [InlineData("%E0%80%AF")]
+        [InlineData("%E0%9F%BF")]
+        [InlineData("%F0%80%80%AF")]
+        [InlineData("%F0%8F%8F%BF")]
+        // Incomplete
+        [InlineData("%")]
+        [InlineData("%%")]
+        [InlineData("%A")]
+        [InlineData("%Y")]
```

Comment:
Look into the `NegativeAssert` you will find it now compare the result to the raw string. For these negative case input equals to output.